mitiq/folding_cirq.py

```diff
@@ -150,9 +199,22 @@ def fold_gates_from_right(circuit: Circuit, stretch: float) -> Circuit:
     Note:
         Folding a single gate adds two gates to the circuit, hence the maximum stretch factor is 3.
     """
+    if not circuit.are_all_measurements_terminal():
+        raise ValueError(f"Input circuit contains intermediate measurements and cannot be folded.")
+
+    add = False
+    if circuit.has_measurements():
+        measurements = _pop_measurements(circuit)
+        for i in range(len(measurements)):
+            measurements[i][0] = -1  # Make sure the moment to insert into is the last in the circuit
+        add = True
```

Comment:
The above code block appears several times. Perhaps it should be a separate function?  
```
measurements = get_measurements(circuit)
folded = Circuit(reversed(reversed_folded_circuit))
if measurements != None:
    _add_measurements(folded, measurements)
```

This could also avoid the need for the `add` boolean

Reply:
Maybe we could rename `add_mesurements` to `append_measurements` and put both the index update `...  measurements[i][0] = -1` and the if condition `if measurements != None:`, in the body of the function `append_measurements`.

Reply:
Yeah since we always want measurements to be terminal, it makes sense to add the block to `_add_measurements` and rename the function to `_append_measurements`.